src/base/TypeChecker.ml

```diff
@@ -167,6 +154,90 @@ module ScillaTypechecker (SR : Rep) (ER : Rep) = struct
     in
     go sctyp [] pattern
 
+  (* tm[tvar := tp]
+     Parallel implementation to the one in Syntax.ml to allow gas accounting.
+  *)
+  let subst_type_in_type_with_gas tvar tp tm =
+    (* Count the number of AST nodes in a type *)
+    let rec type_size t =
+      match t with
+      | PrimType _ | Unit | TypeVar _ -> 1
+      | PolyFun (_, t) -> 1 + type_size t
+      | MapType (t1, t2) | FunType (t1, t2) -> 1 + type_size t1 + type_size t2
+      | ADT (_, ts) ->
+          List.fold_left ts ~init:1 ~f:(fun acc t -> acc + type_size t)
+    in
+
+    let subst_type_cost tvar tm tp_size =
+      match tm with
+      | PrimType _ | Unit
+      | MapType (_, _)
+      | FunType (_, _)
+      | ADT (_, _)
+      | PolyFun (_, _) ->
+          1
+      | TypeVar n -> if String.(n = tvar) then tp_size else 1
+    in
+
+    let tp_size = type_size tp in
+    let rec recurser t =
+      let gas_cost = Stdint.Uint64.of_int @@ subst_type_cost tvar t tp_size in
+      (*        let _ = printf "recursing over %s\n" (pp_typ t) in *)
```

Comment:
Do you intend to keep this comment?

Reply:
This piece of code is moved from TypeUtil. Maybe it's best to remove it. I will.